Implement Author Autocomplete (legacy backend and Search)

rclee91 opened the pull request:
- Add author autocomplete settings for indexer.js in backend
- refactor Search service to work with new indexing
- create new search query for author autocomplete
- add supertest as a dev-dependency to pass ESLint

<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses
Fixes #1417 
also a part of issue #3225 (no new tests are added just yet)

<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [x] **New Feature**: Change which adds functionality

## Description

<!-- Please add a detailed description of what this PR does and why it is needed -->
- In the legacy `backend`, I've added `settings` and `mapping` for the `posts` index to use [edge n-gram](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-edgengram-tokenizer.html) for author autocomplete
- Because we couldn't specify `type` on index creation, I've refactored the `Search` service and tests to work with the new indexing. The current search query is not affected by this new change. `pnpm test search` would pass all tests.
- To get author autocomplete suggestions, I've created a new query out of `Search`, and ways to test it and see it in action is as follows

## Steps to test the PR

- get on my branch, `pnpm install`
- **\*\*Important\*\*** remove `redis` cache `rm -rf redis-data`, this is so the posts can be re-indexed
- build Docker Images
```terminal
 docker-compose --env-file ./config/env.development up --build redis elasticsearch posts traefik search nginx planet
```
- Wait for a bit for ElasticSearch to connect, then `pnpm start`
- Wait a bit for the posts to be indexed
- Can now test `Search` queries at http://localhost/v1/search
- The original Search queries shouldn't be affected. The new query can be tested at  http://localhost/v1/search/authors/autocomplete/
- When `author` is queried, it will return the number of suggested authors, and an array of suggested authors. As the query gets longer, results will be more precise

![image](https://user-images.githubusercontent.com/32626950/159820051-b2e25a34-375d-4e16-b5aa-3b0bfb7bc1ad.png)
![image](https://user-images.githubusercontent.com/32626950/159820068-4636864e-1687-487c-8f1d-784b783a9dd1.png)
![image](https://user-images.githubusercontent.com/32626950/159820178-bd283fb0-495f-4eea-92c5-22ecf0bb04d8.png)
![image](https://user-images.githubusercontent.com/32626950/159820209-2e6f87e3-4445-46a2-be6c-8cab7e0032a2.png)
![image](https://user-images.githubusercontent.com/32626950/159820235-94c09589-6c74-4bf2-982f-74f83668bdc3.png)


<!-- Please add steps to build the changes in your PR locally so that peers can test your PR
    Example:
    - `npm install`
    - Go to '...'
    - Click on '....'
    - Scroll down to '....'
    - See error

    Please remember, the more clear you are, the faster your PR will be reviewed and merged.
 -->
 
## Problems
There is still a lot of tweaking that can be done
- A strict validation for the  http://localhost/v1/search/authors/autocomplete/ isn't yet in place. Currently I am only hijacking the old search validation. If we don't have an `author` query and it passes validation, we'll get a `Error 400`
![image](https://user-images.githubusercontent.com/32626950/159821037-2c2183a3-0ba3-4857-9d5a-3db80ef8afb9.png)
- A query of only one letter won't pass validation, it has to be 2 or more letters. This is part of our rules in `validation.js`
![image](https://user-images.githubusercontent.com/32626950/159821236-b67bad4b-ad56-4091-878b-8eee57f6edee.png)

## Checklist

<!-- Before submitting a PR, address each item -->

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)variables are added/changed or an explanation of why it does not(if applicable)

humphd commented on src/api/search/src/routes/query.js:
```diff
@@ -12,4 +12,13 @@ router.get('/', validateQuery, async (req, res, next) => {
 });
 
+// query for autocomplete
+router.get('/complete', validateQuery, async (req, res, next) => {
```
I wonder if we should do `/authors/autocomplete` so that we can do `/posts/autocomplete` later etc.?

rclee91 replied:
Sure, I can change it to that, cause its more straightforward.
I also think its also possible to do `multi_match` queries on multiple autocomplete fields in the future too, and we can always modify what is returned from the queries. This is just assumption based off of my research, nothing concrete yet.

humphd reviewed (commented):
This is amazing!  I've done one read through, and will spend more time on it tomorrow, but it looks fantastic.

We should figure out if we can migrate this to the parser service, since it's landing soon (can happen in follow-up).

rclee91 commented:
I've changed the route name, and I've also added `highlights`
ElasticSearch can help us wrap the searched terms with `<em>...</em>` tags. Which might be useful for the front end. 
So we can decide if we want to keep one of `authors` and `highlights` or both

(I tested without waiting for all feeds to put in)
![image](https://user-images.githubusercontent.com/32626950/159838003-d3e121b1-5fbc-45a1-be3e-436002ae225b.png)

humphd commented on src/api/search/src/routes/query.js:
```diff
@@ -12,4 +12,13 @@ router.get('/', validateQuery, async (req, res, next) => {
 });
 
+// query for author autocomplete
+router.get('/authors/autocomplete', validateQuery, async (req, res, next) => {
```
I wonder if we need to write another validator, since `validateQuery` is overly broad for what you need here.  That would solve your 2 character limit too.  What about adding `validateAuthor` in https://github.com/Seneca-CDOT/telescope/blob/d8867e9c7f001574ec85630103653dc92a93acfc/src/api/search/src/bin/validation.js.

humphd commented on src/api/search/src/search.js:
```diff
@@ -103,4 +101,57 @@ const search = async (options) => {
+  const {
+    body: { hits },
+  } = await client.search({
+    size: 10000,
```
I wonder if there's an upper-limit on what we should return so it's responsive to use in the front-end?  There's probably some guidance on how many results to return for autocomplete queries so that you have useful but small downloads, and it can query and re-render everything fast enough to be useful when typing.

I wonder if we should consider using 10? See https://ux.stackexchange.com/posts/56131/revisions

rclee91 replied:
I'd like to get a confirmation on this (please see conversation below)
I don't mind changing it to `10` or something else smaller than `10,000`
I just hope that it is clear that if it is too small we might get fewer author results than what was expected of.

humphd replied:
Let's begin with this, and reduce it as we learn how it performs.

humphd commented on src/api/search/src/search.js:
```diff
@@ -103,4 +101,57 @@ const search = async (options) => {
+    body: results,
+  });
+
+  // Filter through all authors and remove duplicates
```
Do we store duplicates?

If so, a faster way to do this is to put the data into a `Set` and then return an Array from that, since Sets don't include duplicates:

```js
const authors = [...new Set(hits.hits)];
```

rclee91 replied:
I'll try to answer the "size", "duplicate", and "highlights" conversation here.
Below is what a response from an ES Search request looks like. 
(I am using sample data on Kibana)
```js
{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : 0.28194553,
    "hits" : [
      {
        "_index" : "posts",
        "_id" : "2",
        "_score" : 0.28194553,
        "_source" : {
          "author" : "Dave Humph"
        },
        "highlight" : {
          "author.autocomplete" : [
            "<em>Da</em>ve <em>H</em>umph"
          ]
        }
      },
      {
        "_index" : "posts",
        "_id" : "3",
        "_score" : 0.28194553,
        "_source" : {
          "author" : "Dave Humph"
        },
        "highlight" : {
          "author.autocomplete" : [
            "<em>Da</em>ve <em>H</em>umph"
          ]
        }
      },
      {
        "_index" : "posts",
        "_id" : "1",
        "_score" : 0.24156076,
        "_source" : {
          "author" : "David Humphrey"
        },
        "highlight" : {
          "author.autocomplete" : [
            "<em>Da</em>vid <em>H</em>umphrey"
          ]
        }
      }
    ]
  }
}
```

The "duplicates" are from the posts, cause each author could have more than one post. 
So if we set the `size` to 10, there's a possibility that the first 10 hits from ElasticSearch are all from the same one author.
Right now its set to the maximum amount of hits that it can page through so we make sure we get all `authors` that match the searched query, and then we take out the duplicates. Currently ElasticSearch doesn't have a way of knowing they are "duplicates", because for ES, they are different results from different "posts".

We won't really have this problem if we were autocompleting on another field with more unique text, for example the `title` of each post. 

Currently the `authors` and `highlights` array are from two different properties (`_score` and `highlight`) in the `hits.hits` array. But, since they're technically the same results, just one with `<em>...</em>` tags and one without, I'm assuming we only want to keep one of them then?

I don't think `Set` will work with `hits.hits` since all the `_id` for each post is different, and so will some `_score`. The `_score` is how ElasticSearch determines how relevant the result is to the query.

humphd replied:
Understood, thanks for clarifying and educating me.

humphd commented on src/api/search/src/search.js:
```diff
@@ -103,4 +101,57 @@ const search = async (options) => {
+    []
+  );
+
+  // Filter through all highlighted authors and remove duplicates
```
Didn't you just remove dupes above?

humphd commented on src/backend/utils/indexer.js:
```diff
@@ -7,7 +7,58 @@ const { client } = require('../lib/elastic');
+  try {
+    const response = await client.indices.exists({ index });
+    if (response.statusCode === 404) {
+      await client.indices.create({
```
Given that you're the only person who understands this code, and we'll have to maintain it, can you add a big comment above this index creation explaining how it works, linking to relevant docs, etc?

TueeNguyen replied:
[The documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-edgengram-tokenizer.html#analysis-edgengram-tokenizer) and [this video](https://www.youtube.com/watch?v=FSPwonNOldk&ab_channel=Codetuber)  helped me understand

Kevan-Y commented:
@rclee91 if I put 1 letter in author, does it suppose to return invalid?
`http://localhost/v1/search/authors/autocomplete/?author=k`
![image](https://user-images.githubusercontent.com/58233223/159962917-ae95ad19-6c0c-4eea-b1cb-2b527d3d1ef2.png)

rclee91 commented on src/backend/utils/indexer.js:
```diff
@@ -7,7 +7,58 @@ const { client } = require('../lib/elastic');
+              },
+            },
+          },
+          mappings: {
+            properties: {
+              author: {
+                type: 'text',
+                fields: {
+                  autocomplete: {
+                    type: 'text',
+                    analyzer: 'autocomplete_analyzer',
+                    search_analyzer: 'autocomplete_search_analyzer',
+                  },
+                },
+                analyzer: 'standard',
+              },
+            },
+          },
```
In theory yes. But I haven't done anything in practise to confirm.

DukeManh commented on src/api/search/src/search.js:
```diff
@@ -103,4 +101,57 @@ const search = async (options) => {
+
+  // Filter through all authors and remove duplicates
+  const authors = hits.hits.reduce(
+    (acc, { _source }) => (acc.includes(_source.author) ? acc : [...acc, _source.author]),
```
```suggestion
    (acc, { _source }) => (acc.includes(_source.author) ? acc : acc.concat(_source.author),
```

rclee91 replied:
I'm just curious if there's a reason to use `concat` over what was before. Better readability?

DukeManh replied:
Readability yeah, why use spread syntax when you can use a built-in function. 
I would personal write `acc.push(_source.auth)` to get O(1) 🤓

rclee91 replied:
Except I think `acc.push` just returns the new length of the array. To use it we can't use it in one line and would still have to `return acc` somewhere in the callback, overall O(n). 

I guess they're all the same. I'll change to `concat` for better readability.

DukeManh replied:
Correct me if im wrong but arrays are JS objects and are passed by reference. So if you push to `acc` and return it, it's not copied(O(n)) but reused in the next iteration.

Kevan-Y replied:
JS object is passed by reference

humphd replied:
None of this matters, we're talking about 10, 100, 1000 iterations?  That's not where this code is going to be slow.  What's slow is the network transfer.  Don't optimize the wrong thing.

DukeManh replied:
I agree, it's premature optimization , won't make much difference. We just discussing for the best solution

DukeManh commented on src/api/search/src/search.js:
```diff
@@ -103,4 +101,57 @@ const search = async (options) => {
+  );
+
+  // Filter through all highlighted authors and remove duplicates
+  const highlights = hits.hits.reduce(
```
Since highlight and author are from the same search result, we could merge them into one array instead of using the same index to get the highlight and author of a result. The object can look like this.
```js
[{
author: 'David Li',
highlight: '"<em>David</em> Li"
},
{...}
]
```

DukeManh replied:
Btw, should we enable fuzzy query in the `/autocomplete` route?

rclee91 replied:
That depends, how accurate we'd like to be with name searches, or how fuzzy we want it to be. 
I personally don't think we need it for just names.

Kevan-Y replied:
> Since highlight and author are from the same search result, we could merge them into one array instead of using the same index to get the highlight and author of a result. The object can look like this.
> 
> ```js
> [{
> author: 'David Li',
> highlight: '"<em>David</em> Li"
> },
> {...}
> ]
> ```

Yeah, agree to merge it, I think it would be easier in the front end to use this data if we want the highlight matching the author.

DukeManh reviewed (changes requested):
Google should consider hiring us.

TueeNguyen reviewed (commented):
~~I had to spend some time learning on Edge n gram which was very interesting.
I tried to test this PR, the server ran but I couldn't to get any result back.~~
~~* `http://localhost/v1/search/complete/?author=Da` returns 404~~
~~* `http://localhost/v1/search/authors/autocomplete/?author=Da` returns 404~~

Edit: my backend took a while to index for some reason

Also the query looks strange, should it be `http://localhost/v1/search/autocomplete?author=Da`

DukeManh commented:
>  tried to test this PR, the server ran but I couldn't to get any result back.

The URL changed to `http://localhost/v1/search/authors/autocomplete/?author=dav`.

Kevan-Y commented:
> ###

I think it would be better to have `http://localhost/v1/search/authors/autocomplete?author=dav`

TueeNguyen commented on src/backend/utils/indexer.js:
```diff
@@ -7,7 +7,58 @@ const { client } = require('../lib/elastic');
+          settings: {
+            analysis: {
+              analyzer: {
+                autocomplete_analyzer: {
```
This is our custom function not a preserve field. (comment)

TueeNguyen commented on src/backend/utils/indexer.js:
```diff
@@ -7,7 +7,58 @@ const { client } = require('../lib/elastic');
+              analyzer: {
+                autocomplete_analyzer: {
+                  tokenizer: 'autocomplete',
+                  filter: ['lowercase', 'remove_duplicates'],
```
If you already have an analyzer that removes duplicate, do you need a filter in `search.js`? Both should be ok.

rclee91 replied:
Its not the same thing. 
This is for removing duplicate `tokens` in the same position when indexing so it doesn't bloat up memory
https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-remove-duplicates-tokenfilter.html

Kevan-Y commented:
> http://localhost/v1/search/authors/autocomplete/?author=dav

That interesting how the routing works, `http://localhost/v1/search/authors/autocomplete/?author=dav` and `http://localhost/v1/search/authors/autocomplete?author=dav` are the same

TueeNguyen commented:
I think whatever is after the `?` counts as query parameter.

rclee91 commented:
Added comments
Added new author validation rule
Merged `author` and `highlight`

humphd commented on src/api/search/src/bin/validation.js:
```diff
@@ -44,14 +44,25 @@ const ValidationRules = [
 const validateQuery = () => {
   return async (req, res, next) => {
-    await Promise.all(ValidationRules.map((rule) => rule.run(req)));
+    const rules = req.route.path === '/' ? ValidationRules : authorRules;
```
This makes me nervous, but let's go with it for now and fix later.  I'd rather split these routes up vs. having this logic determine what we do.

humphd reviewed (approved):
Let's do it.

Kevan-Y reviewed (approved):
LGTM, tested locally it works
